File: 238-product-of-array-except-self/238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        rev, res, suff, pref = [None] * n, [], 1, 1
        
        for idx in range(n-1,-1,-1):
            rev[idx] =  suff
            suff *= nums[idx]
            
        for i in range(n):
            res.append(rev[i] * pref)
            pref *= nums[i]
            
        return res
            

File: 238-product-of-array-except-self/238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> p{1}, res(n);
        for (int i = 0; i < n; i++) {
            p.push_back(p[p.size() - 1] * nums[i]);
        }
        
        int cur = 1;
        for (int i = n- 1; i >= 0; i--) {
            res[i] = p[i] * cur;
            cur *= nums[i];
        }
        return res;
    }
};